File: rpr/*/insert_midi_notes_from_sco-opt.lua
package.path = reaper.GetResourcePath() .. '/Scripts/sockmonkey72 Scripts/MIDI/?.lua'
local mu = require 'MIDIUtils'
mu.CORRECT_OVERLAPS = false

function log(string)
	reaper.ShowConsoleMsg(string .. '\n')
end

function get_tuning_list(take)
	local track = reaper.GetMediaItemTake_Track(take)
	local names, cent_diffs, freqs = {}, {}, {}

	for i = 0, 127 do
		local name, cent_diff, freq = reaper.GetTrackMIDINoteNameEx(0, track, i, 0):match("(%S+)%s+(%S+)%s+(%S+)")
		names[i] = name
		cent_diffs[i] = cent_diff
		freqs[i] = freq
	end

	return names, cent_diffs, freqs
end

function find_index_of_nearest(array, value)
	local min_index = 0
	local min_diff = math.huge

	for i, v in ipairs(array) do
		local diff = math.abs(v - value)
		if diff < min_diff then
			min_diff = diff
			min_index = i
		end
	end

	return array[min_index], min_index
end

function insert_midi_note(take, time, duration, midi_note_num, velocity)
	local ppq_start = reaper.MIDI_GetPPQPosFromProjTime(take, time)
	local ppq_end = reaper.MIDI_GetPPQPosFromProjTime(take, time + duration)

	mu.MIDI_InsertNote(take, true, false, ppq_start, ppq_end, 0, midi_note_num, velocity)

	--reaper.MIDI_InsertNote(take, true, false, ppq_start, ppq_end, 0, midi_note_num, velocity, true)
end



function process_file(file_path, take)
	local names, cent_diffs, freqs = get_tuning_list(take)
	local first_time = nil

	local last_starting_time = 0

	local file = io.open(file_path, 'r')
	if not file then return end

	mu.MIDI_OpenWriteTransaction(take)

	for line in file:lines() do
		
		local values = {}
		for value in line:gmatch('([^,]+)') do
			value = string.gsub(value, "^%s*(.-)%s*$", "%1") -- Remove leading and trailing spaces
			table.insert(values, value)
		end

		--cordelia, 1.342667, 1.000000, 0.090537, 0, 97.945637
		local name, time, dur, dyn, env, freq = table.unpack(values)

		if time ~= last_starting_time then
			if first_time == nil then
				first_time = tonumber(time)
			end
			local nearest_value, midi_note_num = find_index_of_nearest(freqs, tonumber(freq))

			local velocity = math.floor(tonumber(dyn) * 127)

	--[[ 		-- Create the MIDI buffer string for the note event
			local note_on_event = string.pack("BBB", 0x90, midi_note_num, velocity)
			local note_off_event = string.pack("BBB", 0x80, midi_note_num, 0)
			
			-- Calculate the MIDI offset in the buffer
			local offset = math.floor((reaper.MIDI_GetPPQPosFromProjTime(take, tonumber(time))))
			local dur_ppq = math.floor((reaper.MIDI_GetPPQPosFromProjTime(take, tonumber(dur))))
			
			-- Build the MIDI events for note-on and note-off
			local note_on = string.pack("IBs4", offset, 0, note_on_event)
			local note_off = string.pack("IBs4", offset + dur_ppq, 0, note_off_event)
			
			-- Append the events to the buffer
			midi_buffer = midi_buffer .. note_on .. note_off ]]

			insert_midi_note(take, tonumber(time), tonumber(dur), midi_note_num, velocity)
		end
		last_starting_time = time
	end

	mu.MIDI_CommitWriteTransaction(take)

	file:close()
end

function main(file_path)
	local midieditor = reaper.MIDIEditor_GetActive()
	if midieditor ~= nil then
		local take = reaper.MIDIEditor_GetTake(midieditor)
		if take ~= nil then
			process_file(file_path, take)
			reaper.MIDIEditor_OnCommand(midieditor, reaper.NamedCommandLookup('_BR_ME_REFRESH'))
		end
	end
end

local retval, file_path = reaper.GetUserFileNameForRead('', 'Path', 'txt')
if retval then
	--local start_time = reaper.time_precise()
	main(file_path)
	--local end_time = reaper.time_precise()
	--log('Processing time: ' .. tostring(end_time - start_time) .. ' seconds')
end

File: rpr/^/*/insert_midi_notes_from_sco-opt.py
import numpy as np

def log(string):
	RPR_ShowConsoleMsg(str(string) + '\n')

def get_tuning_list(take):
	track = RPR_GetMediaItemTake_Track(take)

	note_data = [str(RPR_GetTrackMIDINoteNameEx(0, track, i, 0)).split() for i in range(0, 127)]
	
	names = [item[0] for item in note_data]
	cent_diffs = [item[1] for item in note_data]
	freqs = [item[2] for item in note_data]

	return names, cent_diffs, freqs

def find_index_of_nearest(array, value):
	array = np.array(array)
	index = np.abs(array - value).argmin()
	nearest_value = array[index]
	return nearest_value, index

def insert_midi_note(take, time: float, duration: float, pitch: int, velocity: int):

	note = {}

	note["selected"] = True  # Set the note as selected
	note["mute"] = False  # Set the note as unmuted
	note["ppq_start"] = RPR_MIDI_GetPPQPosFromProjTime(take, time)  # Set the note start position in PPQ
	note["ppq_end"] = RPR_MIDI_GetPPQPosFromProjTime(take, time + duration)  # Set the note end position in PPQ
	note["pitch"] = pitch  # Set the note pitch (MIDI note number)
	note["vel"] = velocity  # Set the note velocity (0-127)

	# Insert the note into the take
	RPR_MIDI_InsertNote(take, True, False, note["ppq_start"], note["ppq_end"], 0, note["pitch"], note["vel"], True)



def main(file_path):
	# Get the active MIDI editor
	midieditor = RPR_MIDIEditor_GetActive()

	# Check if a MIDI editor is open and the take is valid
	if midieditor != 0 and (take := RPR_MIDIEditor_GetTake(midieditor)) != 0:
		names, cent_diffs, freqs = get_tuning_list(take)
		first_time = None

		with open(file_path, 'r') as file:
			lines = file.read().splitlines()

			for line in lines:
				param = line.strip()
				name, time, dur, dyn, env, freq = param.split(', ')

				if first_time is None:
					first_time = time

				nearest_value, midi_note_num = find_index_of_nearest(freqs, float(freq))

				# round_val = 2
				# main_freq = round(float(freq), round_val)
				# nearest_freq = round(nearest_value, round_val)

				# string = [
				#     'MAIN:\t' + str(main_freq) + ' >> ' + str(nearest_freq) + f' [{names[midi_note_num]}{cent_diffs[midi_note_num]}, {str(midi_note_num)}]',
				#     'BEFORE:\t' + str(round(freqs[midi_note_num-1], round_val)) + f' [{names[midi_note_num-1]}]',
				#     'AFTER:\t\t' + str(round(freqs[midi_note_num+1], round_val)) + f' [{names[midi_note_num+1]}]',
				#     '------------------------------------------------'
				# ]

				# if main_freq != nearest_freq:
				#     log('\n'.join(string))

				insert_midi_note(take, float(time), float(dur), int(midi_note_num), int(float(dyn)*127))


	# Update the MIDI editor display
	RPR_MIDIEditor_OnCommand(midieditor, RPR_NamedCommandLookup("_BR_ME_REFRESH"))

retval, file_path, title, defex = RPR_GetUserFileNameForRead('/Users/j/Documents/PROJECTs/CORDELIA/script', 'Path', 'txt')

if retval:
	main(file_path)
